Workbooks/Azure Active Directory/EntitlementManagementAppRoleAssignments/EntitlementManagementAppRoleAssignments.workbook

```diff
@@ -0,0 +1,105 @@
+{
+  "version": "Notebook/1.0",
+  "items": [
+    {
+      "type": 1,
+      "content": {
+        "json": "## Application role assignment activity"
+      },
+      "name": "text - 2"
+    },
+    {
+      "type": 9,
+      "content": {
+        "version": "KqlParameterItem/1.0",
+        "parameters": [
+          {
+            "id": "a5151809-64d6-45ef-8881-499597a4727c",
+            "version": "KqlParameterItem/1.0",
+            "name": "Since",
+            "label": "Activity since last",
+            "type": 2,
+            "description": "Include audit events since a particular date",
+            "isRequired": true,
+            "value": "30d",
+            "typeSettings": {
+              "additionalResourceOptions": []
+            },
+            "jsonData": " [ {\"value\":\"1d\",\"label\":\"24 hours\"}, {\"value\":\"7d\",\"label\":\"7 days\"}, {\"value\":\"30d\",\"label\":\"30 days\"}, {\"value\":\"90d\",\"label\":\"90 days\"}, { \"value\":\"3653d\", \"label\":\"all\" }]\r\n",
+            "timeContext": {
+              "durationMs": 86400000
+            }
+          },
+          {
+            "id": "19adae3b-644e-462f-a429-9338ce8d6ab0",
+            "version": "KqlParameterItem/1.0",
+            "name": "Appname",
+            "label": "Application name",
+            "type": 2,
+            "isRequired": true,
+            "query": "AuditLogs | where TimeGenerated > ago({Since:value}) and (Category == \"UserManagement\" or Category == \"GroupManagement\" or Category == \"ApplicationManagement\") and Result == \"success\" and (ActivityDisplayName == \"Add app role assignment grant to user\" or ActivityDisplayName == \"Add app role assignment to group\" or ActivityDisplayName == \"Add app role assignment to service principal\" or ActivityDisplayName == \"Add a deletion-marked app role assignment grant to user as part of link removal\" or ActivityDisplayName == \"Add a deletion-marked app role assignment grant to group as part of link removal\" or ActivityDisplayName == \"Add a deletion-marked app role assignment grant to service principal as part of link removal\" or ActivityDisplayName == \"Remove app role assignment from user\" or ActivityDisplayName == \"Remove app role assignment from group\" or ActivityDisplayName == \"Remove app role assignment from service principal\") |project TargetResources | mv-expand TargetResources | extend targetAppName=tostring(TargetResources.displayName) | where targetAppName != \"\" | distinct targetAppName | order by targetAppName asc",
```

Comment:
instead of all these `x == a or x == b or x == c or x == d` you could use the `in` operator instead?
`x in (a, b, c, d)`.  the in syntax usually ends up much smaller since you don't repeat "activitydisplayname ==" all those times.

Reply:
also, do you want this dropdown to return a default value, so that something is selected by default? see 
https://github.com/microsoft/Application-Insights-Workbooks/blob/master/Documentation/Parameters/DropDown.md#parameter-value-label-selection-and-group

that would make it so when the report opens, something is selected and the rest of the report populates with a default set of data?

Reply:
an even more advanced version of this would be to have the activities in a hidden parameter, either a multi-value or multi-select dropdown:
![image](https://user-images.githubusercontent.com/10158007/107709436-10499600-6c7a-11eb-92b0-68a0b1ca7319.png)

which you could then reference in the query text as a parameter instead
```
AuditLogs 
| where TimeGenerated > ago({Since:value}) and Result == "success" 
and (Category in ("UserManagement", "GroupManagement" , "ApplicationManagement") 
and ActivityDisplayName in ({Activities})
| project TargetResources | mv-expand TargetResources | extend targetAppName=tostring(TargetResources.displayName) | where targetAppName != "" | distinct targetAppName | order by targetAppName asc
```

you could do similar with the categories as well.  then advanced users can edit the workbook and see they could do further filtering, or you could make them always available?

Reply:
This could presumably look then something like:
![image](https://user-images.githubusercontent.com/10158007/107711679-dfb72b80-6c7c-11eb-8fe1-f42d8b07d5bc.png)


![image](https://user-images.githubusercontent.com/10158007/107711455-8949ed00-6c7c-11eb-97d9-cb3889de2c40.png)

![image](https://user-images.githubusercontent.com/10158007/107711643-d4fc9680-6c7c-11eb-9d6d-d577d0f7b930.png)

and the queries all can get much simpler if you do that, category and activity become parameters, etc:
```
AuditLogs | where TimeGenerated {Since} and Category in ({Categories}) and Result == "success" and ActivityDisplayName in ({Activities}) and (InitiatedBy.app.displayName != "{ExcludedInitiatorAppName:value}" or InitiatedBy.app.appid != "")...
```

Reply:
Added Activities as a customer-selectable drop-down, and moved to 'in' operator for Category as well.  Chose not to surface Categories at this time as only a few activities would be relevant per category.